### PR TITLE
Implement StringIO#pread

### PR DESCRIPTION
**Pull request by casperisfine**
Both for being closer to real IOs and also because it's a convenient API in multithreaded scenarios.

cc @kddnewton @nobu

### CODE REVIEW

File: ext/stringio/stringio.c [hunk 1]
```diff
@@ -1583,6 +1583,48 @@ strio_read(int argc, VALUE *argv, VALUE self)
...
+ *  call-seq:
+ *    pread(maxlen, offset)             -> string
+ *    pread(maxlen, offset, out_string) -> string
+ *
+ *  See IO#pread.
+ */
+static VALUE
+strio_pread(int argc, VALUE *argv, VALUE self)
+{
+    VALUE rb_len, rb_offset, str;
```
Comment (kou) [inline ext/stringio/stringio.c:1596]: Can we improve `str` name? For example, `rb_buf`, `rb_out`, ...

**Comment by casperisfine** [comment]
@kou I applied your suggestions.

File: ext/stringio/stringio.c [hunk 2]
```diff
@@ -1583,6 +1583,48 @@ strio_read(int argc, VALUE *argv, VALUE self)
...
+    rb_scan_args(argc, argv, "21", &rb_len, &rb_offset, &rb_buf);
+    long len = NUM2LONG(rb_len);
+    long offset = NUM2LONG(rb_offset);
+
+    if (len < 0) {
+        rb_raise(rb_eArgError, "negative string size (or size too big): %ld", len);
+    }
+
+    if (offset < 0) {
+        rb_exc_raise(rb_syserr_new_str(EINVAL, rb_sprintf("pread: Invalid offset argument: %ld", offset)));
```
Comment (kou) [inline ext/stringio/stringio.c:1606]: Can we use `rb_syserr_fail_str()`?

```suggestion
        rb_syserr_fail_str(EINVAL, rb_sprintf("pread: Invalid offset argument: %ld", offset));
```

(How about using `rb_offset`?)

File: ext/stringio/stringio.c [hunk 3]
```diff
@@ -1583,6 +1583,48 @@ strio_read(int argc, VALUE *argv, VALUE self)
...
+static VALUE
+strio_pread(int argc, VALUE *argv, VALUE self)
+{
+    VALUE rb_len, rb_offset, rb_buf;
+    rb_scan_args(argc, argv, "21", &rb_len, &rb_offset, &rb_buf);
+    long len = NUM2LONG(rb_len);
+    long offset = NUM2LONG(rb_offset);
+
+    if (len < 0) {
+        rb_raise(rb_eArgError, "negative string size (or size too big): %ld", len);
```
Comment (kou) [inline ext/stringio/stringio.c:1602]: How about using `rb_len`?

```suggestion
        rb_raise(rb_eArgError, "negative string size (or size too big): %" PRIsVALUE, rb_len);
```

**Review by kou — COMMENTED** [review]
We need a review for JRuby part from @headius before we merge this.

**Comment by casperisfine** [comment]
@kou done.

> We need a review for JRuby part 

No problem. Note that there are pre-existing failures on the JRuby suite, but the added test does pass on JRuby.

**Comment by kou** [comment]
OK.
I've merged this.

**Pull request merged**